src/modules/silentpayments/main_impl.h

```diff
@@ -7,10 +7,296 @@
+    secp256k1_declassify(ctx, &ret, sizeof(ret));
+    if (ret) {
+        secp256k1_scalar_clear(&addend);
+        secp256k1_scalar_clear(&a_sum_scalar);
+        return 0;
+    } else {
+        /* Clear the addend variable as this is no longer needed at this point
+         * and contains secret data. This saves from needing to remember to clear
+         * this variable from multiple places below */
+        secp256k1_scalar_clear(&addend);
+    }
+    /* Compute input_hash = hash(outpoint_L || (a_sum * G)) */
+    secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &A_sum_gej, &a_sum_scalar);
+    secp256k1_ge_set_gej(&A_sum_ge, &A_sum_gej);
+    /* TODO: comment */
+    secp256k1_declassify(ctx, &A_sum_ge, sizeof(A_sum_ge));
+
+    /* Calculate the input hash and tweak a_sum, i.e., a_sum_tweaked = a_sum * input_hash
+     * This should fail if input hash is greater than the curve order, but this is stastically improbable so
+     * we only do a verify_check here.
+     */
+    secp256k1_silentpayments_calculate_input_hash(input_hash, outpoint_smallest36, &A_sum_ge);
+    secp256k1_scalar_set_b32(&input_hash_scalar, input_hash, &overflow);
+    VERIFY_CHECK(!overflow);
+    secp256k1_scalar_mul(&a_sum_scalar, &a_sum_scalar, &input_hash_scalar);
+    /* _recipient_sort sorts the array of recipients in place by their scan public keys (lexicographically).
+     * This ensures that all recipients with the same scan public key are grouped together, as specified in BIP0352.
+     *
+     * More specifically, this ensures `k` is incremented from 0 to the number of requested outputs for each recipient group,
+     * where a recipient group is all addresses with the same scan public key.
+     */
+    secp256k1_silentpayments_recipient_sort(ctx, recipients, n_recipients);
+    last_recipient = *recipients[0];
+    k = 0;
+    for (i = 0; i < n_recipients; i++) {
+        if ((i == 0) || (secp256k1_ec_pubkey_cmp(ctx, &last_recipient.scan_pubkey, &recipients[i]->scan_pubkey) != 0)) {
+            /* If we are on a different scan pubkey, its time to recreate the shared secret and reset k to 0.
+             * It's very unlikely the scan public key is invalid by this point, since this means the caller would
+             * have created the _silentpayments_recipient object incorrectly, but just to be sure we still check that
+             * the public key is valid.
+             */
+            secp256k1_ge pk;
+            if (!secp256k1_pubkey_load(ctx, &pk, &recipients[i]->scan_pubkey)) {
+                secp256k1_scalar_clear(&a_sum_scalar);
+                return 0;
+            }
+            secp256k1_silentpayments_create_shared_secret(ctx, shared_secret, &a_sum_scalar, &pk);
+            k = 0;
+        }
+        if (!secp256k1_silentpayments_create_output_pubkey(ctx, generated_outputs[recipients[i]->index], shared_secret, &recipients[i]->labeled_spend_pubkey, k)) {
+            secp256k1_scalar_clear(&a_sum_scalar);
+            return 0;
```

Comment:
should also memclear the shared secret here and in the error path a few lines above (admitedly a bit annoying though that this leads to so much code duplication :/)